### ISSUE TITLE
JSON_ARR Variable Index

### ISSUE THREAD

**Issue by coltmcnealy-lh** [littlehorse-enterprises/littlehorse#580]
### Context

All other variable types support indexes, but `JSON_ARR` does not.

### Description

It is useful to search a `WfRun` by its variables for two reasons:
- finding + debugging workflows
- using a `WfRun` to store data for a business entity

Consider a `User` workflow (where user:organization is many:many), that has a `JSON_ARR` variable named `organization-ids` containing a list of ID's of `Organization`s to which the `User` belongs. What if we want to find all `User`s belonging to an `Organization`? We would need to do a query on the `organization-ids` variable.

Currently, this is not supported, because properly modeling 

### Acceptance Criteria

We should be able to do something like the following in our Java SDK:
```
WfRunVariable var = wf.addVariable("some-var", VariableType.JSON_ARR)
``` 
If the items of the list are actually json objects, then the following should be valid too:
```
WfRunVariable var = wf.addVariable("some-var", VariableType.JSON_ARR).searchableOn(...)
``` 
We should be able to do the equivalent for the other SDK's.

We should be able to pass in a JSONPath field to query such variables in the `rpc SearchVariable`.

### Out of Scope

Out of scope is editing the `VariableMatch` on the `SearchWfRunRequest` to support `JSON_OBJ` and `JSON_ARR` variables. That will require a new ticket.

Additionally, while defining this ticket it would be prudent to define the way we handle lists in `JSON_OBJ` sub-objects. This is because `$.foo.[].bar` is a valid JSONPath query, and as such we allow creating indexes on that. Currently, the behavior of the server is undefined in the case that a user creates a `JSON_OBJ` index with such a JSONPath.

This ticket does **not** involve fixing the `JSON_OBJ` index, however. That can be done separately.

## Dealing with Multiple Matches
While this ticket seems simple at first, it is actually quite interesting. For example, consider the `SearchVariable` rpc:

```
  rpc SearchVariable(SearchVariableRequest) returns (VariableIdList) {}
```

How should this work when, for example, we have a `JSON_ARR` variable that is just a simple `List<String>`, and we are querying for equality to `"some-value"`. Let's say the `Variable` in question is:

```
["some-value", "another-value", "some-value"]
```

Should the `VariableIdList` returned by the RPC contain the ID of that `Variable` twice? Or only once? Should we extend the `VariableIdList` to include other information, perhaps how many times the query is matched?

My instincts tell me we should go with "the `VariableIdList` should contain the `VariableId` only once, no matter how many times it matches. The semantics of the `SearchVariable` rpc is that it finds the ID's of all `Variable`s that match certain criteria. Matching the criteria is a boolean concept.

Under the hood, this happens to be actually the easiest one to implement because it means that we only need to store one `Tag` for this variable.

## Note on JSONPath

The `Variable` in the above example only contains primitive values (in this case, `STR`). I do not believe we need the `json_path` field here. The `SearchVariableRequest` should just look like:

```
{
  value: {"str": "some-value"},
  wf_spec_name: "my-workflow",
  "var_name": "my-variable"
}
```
And in the `WfSpec` SDK, we probably just need `myVariable.searchable()`.

However, what if the variable looked like:

```
[ {"make": "Ford", "model": "Escape"}, {"make": "Toyota", "model": "Prius"}]
```
and we wanted to search for all `Variable`s containing a Car of `make: Ford`? We would do something like:
```
{
  value: {"str": "Ford"},
  wf_spec_name: "my-workflow",
  "var_name": "cars",
  "json_path": "$.make"
}
```
The Workflow SDK should be like: `carsVar.searchableOn("$.make")`.

### With `JSON_OBJ` Variables

Currently the Workflow SDK allows you to _index_ a `JSON_OBJ` variable, but the `SearchVariable` rpc (and the `SearchWfRun` rpc) don't let you search based on the `JSON_OBJ` variables.

The `json_path` field on the `VariableIndex` should be required when the variable is of type `JSON_OBJ`. That is because (for performance + storage reasons) we can only tag a primitive field. _This contrasts with `JSON_ARR` variables, because in `JSON_ARR` the variable could have primitive types in the root, not requiring a JSONPath._

In `JSON_OBJ`, what if we have a `$.cars[].make` field under index? Then, just like with `JSON_ARR` variables, we can have multiple `Tag`s with the matching values. The semantics of the search should be the same as with `JSON_ARR` too.

## Note on Full Match
Again, the above example only checks for a single element in the list matching some value. We will NOT support full matches in this ticket (i.e. comparing the entirety of the JSON List) because storing the `Tag`'s would be too expensive.

### Components

- [ ] Dashboard
- [ ] Server
- [ ] Python SDK
- [ ] Go SDK
- [ ] Java SDK
- [ ] C# SDK
- [ ] LH Control
- [ ] LH Tests Utils

**Comment by coltmcnealy-lh** [comment]
After a discsussion with @sauljabin and @eduwercamacaro , this ticket is approved to be implemented as discussed. Starting now.

**Comment by coltmcnealy-lh** [comment]
During implementation, it was decided that `JSON_ARR` indexes will have the following properties:

* the `json_path` field is not supported yet.
* Only primitive fields at the root of the list will be indexed.
* All (primitive) fields in the root of the list will be indexed.

This means that only `.searchable()` is supported on a `WfRunVariable` of type `JSON_ARR`. Additionally, the following variable:
```
[1, 2, "foo", [3, 4], {"bar": "baz"}]
```
Will be searchable on:
* 1
* 2
* "foo"
and the sublist and sub-map are ignored by the index.
